File: Python/MigratoryBirds.py
# https://www.hackerrank.com/challenges/migratory-birds/problem

class solution:
    def migratoryBirds(arr):
        freq = {}
        maxCount = 0
        minElement = float('inf')
        
        for bird in arr:
            freq[bird] = freq.get(bird, 0) + 1
            
            if freq[bird] > maxCount or (freq[bird] == maxCount and bird < minElement):
                maxCount = freq[bird]
                minElement = bird
            
        return minElement
    
    if __name__ == '__main__':
        arr = [1, 4, 4, 4, 5, 3]
        print("Answer is", migratoryBirds(arr))

File: MigratoryBirds.java
// https://www.hackerrank.com/challenges/migratory-birds/problem

import java.util.*;

class MigartoryBirds{
    public static int migratoryBirds(List<Integer> arr) {
        int maxCount = 0;
        int element = 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < arr.size(); i++){
            if(!map.containsKey(arr.get(i))){
                map.put(arr.get(i), 1);
            }
            else{
                map.put(arr.get(i), map.get(arr.get(i)) + 1);
            }
            if(map.get(arr.get(i)) > maxCount){
                element = arr.get(i);
                maxCount = map.get(arr.get(i));
            }
            else if(map.get(arr.get(i)) == maxCount && element > arr.get(i)){
                element = arr.get(i);
            }
        }
        return element;      
    }
    public static void main(String[] args) {
        List<Integer> arr = Arrays.asList(1, 4, 4, 4, 5, 3);
        System.out.println("Most Frequent Bird Type: " + migratoryBirds(arr));
    }
}